src/engine/GroupByImpl.cpp

```diff
@@ -31,6 +31,212 @@
+    using LazyResult = Result::LazyResult;
+    using IdTableVocabPair = Result::IdTableVocabPair;
+    // No need for final commit when loop was never entered.
+    if (!groupSplitAcrossTables_) {
+      // If we have an implicit group by we need to produce one result row
+      if (groupByCols_.empty()) {
+        finalValue_ = LazyResult(lazySingleValueRange([this]() {
+          // If we have an implicit GROUP BY, where the entire input is a
+          // single group, we need to produce one result row.
+          parent_->processEmptyImplicitGroup<OUT_WIDTH>(
+              resultTable_, aggregates_, &currentLocalVocab_);
+          return IdTableVocabPair{std::move(resultTable_),
+                                  std::move(currentLocalVocab_)};
+        }));
+
+      } else if (singleIdTable_) {
+        // Yield at least a single empty table if requested.
+        finalValue_ = LazyResult(lazySingleValueRange([this]() {
+          return IdTableVocabPair{std::move(resultTable_),
+                                  std::move(currentLocalVocab_)};
+        }));
+      } else {
+        // no final values to add
+        isFinished_ = true;
+      }
+      return;
+    }
+
+    // Process remaining items in the last group.  For those we have already
+    // called `lazyGroupBy.processBlock()` but the call to `commitRow` is
+    // still missing. We have to setup a dummy input table and evaluation
+    // context, that have the values of the `currentGroupBlock` in the
+    // correct columns.
+    finalValue_ = LazyResult(lazySingleValueRange([this]() {
+      IdTable idTable{inWidth_, ad_utility::makeAllocatorWithLimit<Id>(
+                                    1_B * sizeof(Id) * inWidth_)};
+      idTable.emplace_back();
+      for (const auto& [colIdx, value] : currentGroupBlock_) {
+        idTable.at(0, colIdx) = value;
+      }
+
+      sparqlExpression::EvaluationContext evaluationContext =
+          parent_->createEvaluationContext(currentLocalVocab_, idTable);
+
+      lazyGroupBy_->commitRow(resultTable_, evaluationContext,
+                              currentGroupBlock_);
+      currentLocalVocab_.mergeWith(storedLocalVocabs_);
+      return IdTableVocabPair{std::move(resultTable_),
+                              std::move(currentLocalVocab_)};
+    }));
+  }
+
+  void yieldFinalValue() {
+    buildFinalValue();
+    if (!finalValue_.has_value()) {
+      isFinished_ = true;
+      return;
+    }
+
+    // calling begin ensures that the final value is ready
+    it_ = ql::ranges::begin(finalValue_.value());
+  }
+
+  void start() {
+    using namespace ad_utility;
+    lazyGroupBy_ = std::make_unique<LazyGroupBy>(
+        currentLocalVocab_, std::move(aggregateAliases_),
+        parent_->getExecutionContext()->getAllocator(), groupByCols_.size());
+
+    range_ = Result::LazyResult(CachingContinuableTransformInputRange(
+        subresult_->idTables(),
+        [this](Result::IdTableVocabPair& pair) { return process(pair); }));
```

Comment:
`lazyGroupBy_` and `range_` should be initialisable in the constructor of this class. No need to wait until here.

Reply:
Stuff like this can also happen if the initialisation order is wrong, but not that important, your solution seems reasonable enough.